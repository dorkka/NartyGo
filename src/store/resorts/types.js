export const SET_RESORTS = 'resorts/SET_RESORTS';
export const IS_LOADING = 'resorts/IS_LOADING';
export const SET_ERROR = 'resorts/SET_ERROR';
export const SET_RESORT = 'resorts/SET_RESORT';
export const SET_SELECTED_OPTIONS = 'SET_SELECTED_OPTIONS';
