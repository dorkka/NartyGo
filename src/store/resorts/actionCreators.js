import * as types from './types';
import resourceFetcher from '../../services/resourceFetcher';

export const setResorts = (data, headers, perPage) => {
  const pageCount = Math.ceil(headers.get('x-total-count') / perPage);
  const list = [];
  const byId = {};
  data.forEach(element => {
    list.push(element.id);
    byId[element.id] = element;
  });
  return {
    type: types.SET_RESORTS,
    payload: {
      pageCount,
      list,
      byId,
    },
  };
};

export const setError = (error) => ({
  type: types.SET_ERROR,
  error,
});

export const setIsLoading = () => ({
  type: types.IS_LOADING,
});

export const setSpecificResort = data => {
  const byId = { [data.id]: data };
  return {
    type: types.SET_RESORT,
    payload: {
      byId,
    },
  };
};

export const getResorts = (_page, _limit) => (dispatch) => {
  dispatch(setIsLoading());
  return resourceFetcher('http://localhost:3001/resorts')({ _page, _limit })
    .then(({ data, headers }) => {
      dispatch(setResorts(data, headers, _limit));
    })
    .catch(error => dispatch(setError(error)));
};

export const getSpecificResort = (id) => (dispatch, getState) => {
  if ((getState().resorts.byId[id])) {
    return null;
  }
  dispatch(setIsLoading());
  return resourceFetcher(`http://localhost:3001/resorts/${id}`)()
    .then(({ data }) => {
      dispatch(setSpecificResort(data));
    })
    .catch(error => dispatch(setError(error)));
};

export const setSelectOptions = (data) => {
  const resortsOptions = []
  const cityOptions = []
  let label = ''
  data.forEach((element) => {
    label = `${element.name} - ${element.city}`;
    resortsOptions.push({ value: element.name, label, id: element.id })
    cityOptions.push({ value: element.city, label, id: element.id })
  })
  return {
    type: types.SET_SELECTED_OPTIONS,
    payload: {
      resortsOptions,
      cityOptions,
    }
  }
}

export const getAllResorts = () => (dispatch) => {
  resourceFetcher('http://localhost:3001/resorts')().then(({ data }) => {
    dispatch(setSelectOptions(data))
  });
}
