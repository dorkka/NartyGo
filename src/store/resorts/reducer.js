import * as types from './types';

const initialState = {
  byId: {},
  list: [],
  pageCount: 1,
  isLoading: false,
  error: null,
};

function resortsReducer(state = initialState, action) {
  switch (action.type) {
    case types.SET_RESORTS: {
      return {
        ...state,
        byId: {
          ...state.byId,
          ...action.payload.byId,
        },
        list: action.payload.list,
        pageCount: action.payload.pageCount,
        isLoading: false,
      };
    }
    case types.SET_ERROR: {
      return { ...state, error: action.payload.error };
    }
    case types.IS_LOADING: {
      return { ...state, isLoading: true };
    }
    default: {
      return state;
    }
  }
}
export default resortsReducer;
