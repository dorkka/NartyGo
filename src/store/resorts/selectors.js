export const getCurrentResorts = (state) =>
  state.resorts.list.map((id) => state.resorts.byId[id] || {});
export const getSpecificResort = (state, id) =>
  state.resorts.byId[id] || { weather: {}, piste: [] };
