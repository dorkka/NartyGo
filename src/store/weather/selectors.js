import moment from 'moment';

export const getCurrentWeather = (state, cityId) => {
  const weatherFromStore = state.weather.byId[cityId] || {};
  const weatherWithDefaults = {
    main: {}, weather: [{}], wind: {}, sys: {}, ...weatherFromStore,
  };
  return ({
    temperature: weatherWithDefaults.main.temp,
    temperatureMin: weatherWithDefaults.main.temp_min,
    temperatureMax: weatherWithDefaults.main.temp_max,
    pressure: weatherWithDefaults.main.pressure,
    wind: weatherWithDefaults.wind.speed,
    clouds: weatherWithDefaults.weather[0].description,
    sunrise: moment(weatherWithDefaults.sys.sunrise).format('h:mm:ss'),
    sunset: moment(weatherWithDefaults.sys.sunset).format('h:mm:ss'),
  });
};

