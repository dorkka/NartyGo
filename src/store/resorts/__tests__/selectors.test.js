import * as selectors from '../selectors';
import { resorts } from '../../../specs/fixtures/resorts';

describe('selectors', () => {
  const state = {
    resorts: {
      byId: {
        [resorts[0].id]: resorts[0],
        [resorts[1].id]: resorts[1],
      },
      list: [1, 2],
      pageCount: 1,
      isLoading: false,
      error: null,
    },
  };
  const intialState = {
    resorts: {
      byId: {},
      list: [],
      pageCount: 1,
      isLoading: false,
      error: null,
    },
  };

  describe('getCurrentResort', () => {
    test('should return current resorts when resorts in state', () => {
      expect(selectors.getCurrentResorts(state)).toMatchSnapshot();
    });
    test('should return empty array when no resots in state', () => {
      expect(selectors.getCurrentResorts(intialState)).toEqual([]);
    });
  });

  describe('getSpecificResort', () => {
    test('should return resort when it is in state', () => {
      expect(selectors.getSpecificResort(state, 2)).toMatchSnapshot();
    });
    test('should return empty object when no resort in state', () => {
      expect(selectors.getSpecificResort(state, 6)).toEqual({ weather: {}, piste: [] });
    });
  });
});

