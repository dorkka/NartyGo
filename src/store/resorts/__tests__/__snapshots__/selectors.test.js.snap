// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors getCurrentResort should return current resorts when resorts in state 1`] = `
Array [
  Object {
    "city": "Ustroń",
    "cityId": 7531846,
    "coordinates": Object {
      "lat": 49.6833162,
      "lng": 18.8196089,
    },
    "id": 1,
    "name": "Czantoria",
    "piste": Array [
      Object {
        "id": 1,
        "name": "red",
        "number": 1,
      },
      Object {
        "id": 2,
        "name": "blue",
        "number": 2,
      },
    ],
  },
  Object {
    "city": "Karlików",
    "cityId": 7531370,
    "coordinates": Object {
      "lat": 49.43529,
      "lng": 22.067767,
    },
    "id": 2,
    "name": "Karlików",
    "piste": Array [
      Object {
        "id": 1,
        "name": "Mała Palenica",
        "number": 1,
      },
    ],
  },
]
`;

exports[`selectors getSpecificResort should return resort when it is in state 1`] = `
Object {
  "city": "Karlików",
  "cityId": 7531370,
  "coordinates": Object {
    "lat": 49.43529,
    "lng": 22.067767,
  },
  "id": 2,
  "name": "Karlików",
  "piste": Array [
    Object {
      "id": 1,
      "name": "Mała Palenica",
      "number": 1,
    },
  ],
}
`;
